components/eamxx/src/physics/dp/impl/dp_iop_setinitial_impl.hpp

```diff
@@ -12,11 +12,203 @@ namespace dp {
+      policy,
+      KOKKOS_LAMBDA(const MemberType& team) {
+
+      // We cannot support parallelism at the element level when nstep<=1 because the thelev
+      // computation of tobs is dependent on prior iterations that may have alterted tobs
+      for (Int ie = 0; ie < nelemd; ++ie) {
+        for (Int j = 0; j < np; ++j) {
+          for (Int i = 0; i < np; ++i) {
+
+          // Find level where tobs is no longer zero
+          Int thelev=-1;
+          if (nstep <= 1) {
+            Kokkos::parallel_reduce(
+              Kokkos::TeamVectorRange(team, plev_packs), [&] (int plev_pack, Int& pmin) {
+              auto zmask = tobs(plev_pack) != 0;
+              if (zmask.any()) {
+                pmin = plev_pack;
+              }
+            }, Kokkos::Min<Int>(thelev));
+          }
+
+          if (nstep <= 1) {
+            Kokkos::parallel_for(
+              Kokkos::TeamVectorRange(team, thelev+1), [&] (int k) {
+              auto zmask = k < thelev ? Smask(true) : tobs(k) == 0;
+              vector_simd for (Int p = 0; p < Spack::n; ++p) {
+                if (zmask[p]) {
+                  tobs(k)[p] = elem.m_forcing.m_ft(ie,i,j,k)[p];
+                  qobs(k)[p] = tracers.Q(ie,0,i,j,k)[p]; // Tracer index 0 is qobs
+                }
+              }
+            });
+          }
+          else {
+            Kokkos::parallel_for(
+              Kokkos::TeamVectorRange(team, plev_packs), [&] (int k) {
+              // Ekat packs don't know how to assign themselves to KokkosKernels::Batched::Experimental::SIMD
+              vector_simd for (Int p = 0; p < Spack::n; ++p) {
+                tobs(k)[p] = elem.m_forcing.m_ft(ie,i,j,k)[p];
+                qobs(k)[p] = tracers.Q(ie,0,i,j,k)[p];
+              }
+            });
+          }
+
+          if (nstep == 0) {
+            if (scm_zero_non_iop_tracers) {
+              for (Int cix = 0; cix < pcnst; ++cix) {
+                Kokkos::parallel_for(
+                  Kokkos::TeamVectorRange(team, plev_packs), [&] (int k) {
+                  tracers.Q(ie, cix, i, j, k) = qmin(cix);
+                });
+              }
+            }
+
+            Kokkos::parallel_for(
+              Kokkos::TeamVectorRange(team, thelev, plev_packs), [&] (int k) {
+              auto zmask = k > thelev ? Smask(true) : tobs(k) != 0;
+              if (have_t) {
+                vector_simd for (Int p = 0; p < Spack::n; ++p) {
+                  if (zmask[p]) {
+                    elem.m_forcing.m_ft(ie,i,j,k)[p] = tobs(k)[p];
+                  }
+                }
+              }
+              if (have_q) {
+                vector_simd for (Int p = 0; p < Spack::n; ++p) {
+                  if (zmask[p]) {
+                    tracers.Q(ie,0,i,j,k)[p] = qobs(k)[p];
+                  }
+                }
+              }
+            });
+
+            if (have_ps) {
+              elem.m_state.m_ps_v(ie, 0, i, j) = psobs; // what is [NUM_TIME_LEVELS]?
```

Comment:
How to handle the NUM_TIME_LEVELS dimension? f90 didn't use it as far as I can tell.

Reply:
F90 does have it too. It's called `timelevels`, declared in `src/theta-l_kokkos/element_state.F90`.

Reply:
The time levels are used to store copies of the state that are used during the timestepping. For the currently used DIRK method, a 5 stages method, turns out that 3 vectors are enough to store all stages (since you can recycle old stages as the RK method progresses).

Reply:
I think DP probably just wants the most recent (up-to-date) value, right? Is that the last value, or is there a secret decoder ring needed to figure out the order of the time levels?

Reply:
Homme stores the index for the last value in the `TimeLevel` data structure. The index of the most up to date solution should be stored in the `np1` member of `TimeLevel`. However, you need to be careful _when_ you get that. If you get it _before_ `TimeLevel::update_dynamics_levels` has been called, then yes, `np1` is the most recent one. If `update_dynamics_levels` has already been called, then `n0` is the most recent. The update method simply does `np1=nm1`, `n0=np1`, `nm1 = n0`.

If you let `prim_run_subcycle` complete, then the most up to date time slice index is stored in `n0` (since update levels is called toward the end of prim_run_subcycle).

Reply:
I *think* this is called directly before `prim_run_subcycle` in eam.

Reply:
Then `n0` shold be ok.